pallets/rps/src/lib.rs

```diff
@@ -14,89 +13,427 @@ mod tests;
+		ChallengeNotFound,
+		/// The challenge is not open, so it cannot be entered
+		ChallengeNotOpen,
+		/// Cannot accept challenges created by the same account
+		CannotChallengeOneself,
+		/// Cannot play in a challenge in which the account is not a participant
+		CannotPlayInNonParticipatingChallenge,
+		/// Cannot play in a challenge not in the 'Accepted' state
+		ChallengeStateForbidsPlay,
+		/// Cannot play again in an already played challenge
+		ChallengeAlreadyPlayed,
+		/// Challenge state is not Accepted so it cannot be resolved
+		ChallengeStateForbidsResolution,
+		/// The challenge logic reached an invalid state
+		InvalidState,
+		/// Cannot reveal a challenge in which the account is not a participant
+		CannotRevealNonParticipatingChallenge,
+		/// The hash of the original account play and the value indicated in the reveal don't match
+		InvalidHandHash,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(10_000)]
+		pub fn create_challenge(origin: OriginFor<T>, bet_amount: BalanceOf<T>) -> DispatchResult {
+			let challenger = ensure_signed(origin)?;
+
+			let min_amount = T::MinBetAmount::get();
+			ensure!(bet_amount >= min_amount, Error::<T>::InsufficientBetAmount);
+
+			let challenge_id = NextBetId::<T>::get();
+			let challenge_state =
+				ChallengeState::Open(OpenChallenge { challenger: challenger.clone(), bet_amount });
+
+			NextBetId::<T>::mutate(|x| *x += 1);
+
+			ChallengeStore::<T>::insert(&challenge_id, challenge_state);
+
+			Self::deposit_event(Event::ChallengeCreated(challenge_id, challenger, bet_amount));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(10_000)]
+		pub fn enter_challenge(origin: OriginFor<T>, challenge_id: ChallengeId) -> DispatchResult {
+			let rival = ensure_signed(origin)?;
+
+			if let Err(e) = ChallengeStore::<T>::try_mutate(&challenge_id, |challenge_entry| {
+				ensure!(challenge_entry.is_some(), Error::<T>::ChallengeNotFound);
+
+				let challenge_state = challenge_entry.as_mut().unwrap();
+
+				if let ChallengeState::Open(open_state) = challenge_state {
+					if open_state.challenger == rival {
+						Err(Error::<T>::CannotChallengeOneself)
+					} else {
+						*challenge_state = ChallengeState::Accepted(AcceptedChallenge::from_open(
+							open_state.clone(),
+							rival.clone(),
+						));
+						Self::deposit_event(Event::EnteredChallenge(challenge_id, rival));
+						Ok(())
+					}
+				} else {
+					Err(Error::<T>::ChallengeNotOpen)
+				}
+			}) {
+				Err(e.into())
+			} else {
+				Ok(().into())
+			}
+		}
+
+		// play
+		// - Bet Id
+		// - Participant Id
+		// - Hand Payload (Encrypted<(HandType (Rock|Paper|Scissor))>, Public-Key)
+		#[pallet::weight(10_000)]
+		pub fn play_challenge(
+			origin: OriginFor<T>,
+			challenge_id: ChallengeId,
+			challenge_play: ChallengePlay,
+			challenger_secret: u64,
+		) -> DispatchResult {
+			let player = ensure_signed(origin)?;
+
+			if let Some(challenge) = ChallengeStore::<T>::get(&challenge_id) {
```

Comment:
Here it's more idiomatic to write:
```suggestion
			let challenge = ChallengeStore::<T>::get(&challenge_id).ok_or(Error::<T>::ChallengeNotFound)?;
```